### ISSUE TITLE
Cannot change AssemblyName property on command line

### ISSUE THREAD

**Issue by mladedav** [microsoft/msbuild#4696]
From what I found out, -p properties are set recuresively in referenced projects. AssemblyName also changes ProjectId for all referenced projects, which makes restore fail and although build may still suceed, running the project ends in unknown types, because the library cannot be found.

### Steps to reproduce

Command line
```
dotnet new console -n Project1
dotnet new classlib -n Project2
dotnet add Project1 reference Project2
dotnet msbuild -r -p:AssemblyName=foo Project1

```
### Expected  behavior

The project compiles normally.

### Actual behavior

```
C:\Program Files\dotnet\sdk\2.2.401\NuGet.targets(123,5): error : Ambiguous project name 'foo'. [C:\Users\David\Documents\Sandbox\MsBuild\Project1\Project1.csproj]

```

### Environment data
`dotnet msbuild /version` output:
```
Microsoft (R) Build Engine version 16.2.32702+c4012a063 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

16.200.19.32702
```
OS info:
Windows 10
```
Major  Minor  Build  Revision
-----  -----  -----  --------
10     0      18362  0
```
If applicable, version of the tool that invokes MSBuild (Visual Studio, dotnet CLI, etc):
dotnet

**Comment by rainersigwald** [comment]
> -p properties are set recuresively in referenced projects

That's correct; doing this on the command line sets a [global property](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-properties?view=vs-2019#global-properties), which is inherited by all referenced projects by default. It's possible to set metadata on the `ProjectReference` items to tell the `MSBuild` task to avoid passing a specific property down, but I would not recommend making such a change to your projects for this problem.

Can you elaborate about why you want to do this? There may be a different way to accomplish your goal.

**Comment by mladedav** [comment]
We would like to automatically build a dotnet core application and create a docker image around it. We need the name of the built assembly to put it in the CMD in docker. Unfortunately the AssemblyName is created with variables from different files and parsing it does not seem like an easy task. I had hoped to change just the AssemblyName for the entrypoint library.

The solution is built locally (because of a private NuGet feed), so the property cannot be changed before the build in docker. The whole solution could be potentially copied locally, but doing that just to change a single property in a single project file does not seem like a good solution to me either).

Currently we parse the output of dotnet publish to obtain the dll name, but again, I had hoped there would be a better approach.

Thank you for mentioning the global properties, I think I understand the issue now.

**Comment by mohite-abhi** [comment]
Hi did you find any solution for this? I have a similar need of dotnet project automation and can't figure out how to find resulting dll file name. If I try using property injection, it affect the assemblyname at solution level and messes up the build. Please share your approach if you remember it. Thanks in advance.

**Comment by mladedav** [comment]
we ended up partly parsing stdout from dotnet publish.

**Comment by baronfel** [comment]
What I would suggest that folks do is create a target that would run after the `Publish` target, then use the value of the already-computed MSBuild properties for the AssemblyName (and perhaps others like `TargetFile`, etc) to accurately track any changes to the project names and outputs.  You can see an example of this in the [dotnet/sdk-container-builds](https://github.com/dotnet/sdk-container-builds/blob/main/packaging/build/Microsoft.NET.Build.Containers.targets#L72-L77) repo, where we package projects into containers after publishing doing just this process.
